src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.Transform.cs

```diff
@@ -0,0 +1,190 @@
+                }
+
+                break;
+
+            case JsonValueKind.True:
+                if (transformOptions.ConvertBooleanSchemas)
+                {
+                    schema = new JsonObject();
+                }
+
+                break;
+
+            case JsonValueKind.Object:
+                JsonObject schemaObj = (JsonObject)schema;
+                JsonObject? properties = null;
+
+                // Step 1. Recursively apply transformations to any nested schemas we might be able to detect.
+                if (schemaObj.TryGetPropertyValue(PropertiesPropertyName, out JsonNode? props) && props is JsonObject propsObj)
+                {
+                    properties = propsObj;
+                    path?.Add(PropertiesPropertyName);
+                    foreach (var prop in properties.ToArray())
+                    {
+                        path?.Add(prop.Key);
+                        properties[prop.Key] = TransformSchemaCore(prop.Value, transformOptions, path);
+                        path?.RemoveAt(path.Count - 1);
+                    }
+
+                    path?.RemoveAt(path.Count - 1);
+                }
+
+                if (schemaObj.TryGetPropertyValue(ItemsPropertyName, out JsonNode? itemsSchema))
+                {
+                    path?.Add(ItemsPropertyName);
+                    schemaObj[ItemsPropertyName] = TransformSchemaCore(itemsSchema, transformOptions, path);
+                    path?.RemoveAt(path.Count - 1);
+                }
+
+                if (schemaObj.TryGetPropertyValue(AdditionalPropertiesPropertyName, out JsonNode? additionalProps) &&
+                    additionalProps?.GetValueKind() is not JsonValueKind.False)
+                {
+                    path?.Add(AdditionalPropertiesPropertyName);
+                    schemaObj[AdditionalPropertiesPropertyName] = TransformSchemaCore(additionalProps, transformOptions, path);
+                    path?.RemoveAt(path.Count - 1);
+                }
+
+                if (schemaObj.TryGetPropertyValue(NotPropertyName, out JsonNode? notSchema))
+                {
+                    path?.Add(NotPropertyName);
+                    schemaObj[NotPropertyName] = TransformSchemaCore(notSchema, transformOptions, path);
+                    path?.RemoveAt(path.Count - 1);
+                }
+
+                // Traverse keywords that contain arrays of schemas
+                ReadOnlySpan<string> combinatorKeywords = ["anyOf", "oneOf", "allOf"];
+                foreach (string combinatorKeyword in combinatorKeywords)
+                {
+                    if (schemaObj.TryGetPropertyValue(combinatorKeyword, out JsonNode? combinatorSchema) && combinatorSchema is JsonArray combinatorArray)
+                    {
+                        path?.Add(combinatorKeyword);
+                        for (int i = 0; i < combinatorArray.Count; i++)
+                        {
+                            path?.Add($"[{i}]");
```

Comment:
This seems very allocatey. Is it only used rarely?

Reply:
The more `anyOf`, `oneOf`, or `allOf` keywords in the schema the more this will end up being used. I could try doing something smart here like precomputing an array of index strings < 10 but any `JsonNode` transformation is allocatey by definition so I doubt it will have meaningful impact.